### ISSUE TITLE
[SDK 49 beta] SyntaxError: Invalid or unexpected token with yarn berry 

### ISSUE THREAD

**Issue by jmatsushita** [expo/expo#23194]
### Summary

I'm trying out `49.0.0-beta.0` and the "bin/cli" script doesn't work on yarn berry because [of this issue](https://github.com/yarnpkg/berry/issues/882) the error is
```
/node_modules/expo/bin/cli:3
# Forward all arguments to the local CLI (@expo/cli).
^

SyntaxError: Invalid or unexpected token
    at internalCompileFunction (node:internal/vm:73:18)
    at wrapSafe (node:internal/modules/cjs/loader:1176:20)
    at Module._compile (node:internal/modules/cjs/loader:1218:27)
    at Module._extensions..js (node:internal/modules/cjs/loader:1308:10)
    at Module.load (node:internal/modules/cjs/loader:1117:32)
    at Module._load (node:internal/modules/cjs/loader:958:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at node:internal/main/run_main_module:23:47
```

Working around this in a regular repo works with `./node_modules/expo/bin/cli start --web`

### What platform(s) does this occur on?

Web

### SDK Version

49.0.0-beta.0

### Environment
```
  expo-env-info 1.0.5 environment info:
    System:
      OS: macOS 13.4.1
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 18.16.0 - /private/var/folders/38/trffdc5j1p7f26b7809_hfvh0000gn/T/xfs-aa499401/node
      Yarn: 4.0.0-rc.46 - /private/var/folders/38/trffdc5j1p7f26b7809_hfvh0000gn/T/xfs-aa499401/yarn
      npm: 8.19.4 - /nix/store/2js4vxscm74iwzgs3mmqpqpg2kkgwdfg-nodejs-16.20.0/bin/npm
      Watchman: 2023.06.26.00 - /opt/homebrew/bin/watchman
    SDKs:
      iOS SDK:
        Platforms: DriverKit 22.4, iOS 16.4, macOS 13.3, tvOS 16.4, watchOS 9.4
    IDEs:
      Android Studio: 2022.2 AI-222.4459.24.2221.10121639
      Xcode: 14.3.1/14E300c - /nix/store/8m3n01ciipnip4nsklzgh70cryms3m34-xcode-wrapper-14.3.1/bin/xcodebuild
    Expo Workflow: managed
```

### Minimal reproducible example

In the starter repo with `yarn` installed.
```
yarn set berry
yarn start
```

**Comment by vbylen** [comment]
Can confirm, I'm also experiencing this after upgrading to `49.0.0-beta.0`

**Comment by jmatsushita** [comment]
Note that there was a fix [merged](https://github.com/yarnpkg/berry/pull/5508)  last week and I tried it with `yarn set version 4.0.0-rc.47` but  unfortunately it doesn't solve it.

**Comment by jmatsushita** [comment]
I confirm that if I `cd` into the workspace and run `expo` it works, but unfortunately running `eas build --local` after `cd` fails with:
```
[PREBUILD] /private/var/folders/38/trffdc5j1p7f26b7809_hfvh0000gn/T/eas-build-local-nodejs/cef41d77-bb8f-406f-8f48-c40d931cc20f/build/node_modules/expo/bin/cli:3
[PREBUILD] # Forward all arguments to the local CLI (@expo/cli).
[PREBUILD] ^
[PREBUILD]
[PREBUILD] SyntaxError: Invalid or unexpected token
[PREBUILD]     at internalCompileFunction (node:internal/vm:73:18)
[PREBUILD]     at wrapSafe (node:internal/modules/cjs/loader:1176:20)
[PREBUILD]     at Module._compile (node:internal/modules/cjs/loader:1218:27)
[PREBUILD]     at Module._extensions..js (node:internal/modules/cjs/loader:1308:10)
[PREBUILD]     at Module.load (node:internal/modules/cjs/loader:1117:32)
[PREBUILD]     at Module._load (node:internal/modules/cjs/loader:958:12)
[PREBUILD]     at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
[PREBUILD]     at node:internal/main/run_main_module:23:47
[PREBUILD]
[PREBUILD] Node.js v18.16.0
[PREBUILD]
Error: yarn expo prebuild --no-install --platform android exited with non-zero code: 1
    at ChildProcess.completionListener (/Users/jun/.npm/_npx/0019a18d7f70f1a8/node_modules/@expo/spawn-async/build/spawnAsync.js:42:23)
    at Object.onceWrapper (node:events:628:26)
    at ChildProcess.emit (node:events:513:28)
    at maybeClose (node:internal/child_process:1091:16)
    at Socket.<anonymous> (node:internal/child_process:449:11)
    at Socket.emit (node:events:513:28)
    at Pipe.<anonymous> (node:net:322:12)
    ...
    at spawnAsync (/Users/jun/.npm/_npx/0019a18d7f70f1a8/node_modules/@expo/spawn-async/build/spawnAsync.js:7:23)
    at spawn (/Users/jun/.npm/_npx/0019a18d7f70f1a8/node_modules/@expo/turtle-spawn/dist/index.js:16:47)
    at runExpoCliCommand (/Users/jun/.npm/_npx/0019a18d7f70f1a8/node_modules/@expo/build-tools/dist/utils/project.js:31:47)
    at prebuildAsync (/Users/jun/.npm/_npx/0019a18d7f70f1a8/node_modules/@expo/build-tools/dist/common/prebuild.js:25:43)
    at async /Users/jun/.npm/_npx/0019a18d7f70f1a8/node_modules/@expo/build-tools/dist/builders/android.js:48:9
    at async BuildContext.runBuildPhase (/Users/jun/.npm/_npx/0019a18d7f70f1a8/node_modules/@expo/build-tools/dist/context.js:86:28)
    at async buildAsync (/Users/jun/.npm/_npx/0019a18d7f70f1a8/node_modules/@expo/build-tools/dist/builders/android.js:42:5)
    at async runBuilderWithHooksAsync (/Users/jun/.npm/_npx/0019a18d7f70f1a8/node_modules/@expo/build-tools/dist/builders/common.js:13:13)
    at async Object.androidBuilder (/Users/jun/.npm/_npx/0019a18d7f70f1a8/node_modules/@expo/build-tools/dist/builders/android.js:21:16)
    at async buildAndroidAsync (/Users/jun/.npm/_npx/0019a18d7f70f1a8/node_modules/eas-cli-local-build-plugin/dist/android.js:66:12)
```

**Comment by arcanis** [comment]
That's because 49.0.0-beta.0 ships `bin/cli` as a shellscript, which isn't supported as they are non-portable (we really suggest to use JS scripts as entry points, except on native packages where using a native binary is ok provided the `cpu` and `os` fields are properly set in the `package.json`).

Highlighting why shellscripts are a good idea for portability, we can see that this one is even hardcoding npm, so it'd fail in other ways even if Yarn did support that ...

```
#!/usr/bin/env bash

# Forward all arguments to the local CLI (@expo/cli).
npm exec --no-install -- expo-internal "$@"
```

**Comment by jmatsushita** [comment]
Here's a patch which seems to fix the problem for me:
```diff
diff --git a/bin/cli b/bin/cli
deleted file mode 100755
index 200f9b97c40880eb9f83a11358f2dc18b6d478bc..0000000000000000000000000000000000000000
--- a/bin/cli
+++ /dev/null
@@ -1,4 +0,0 @@
-#!/usr/bin/env bash
-
-# Forward all arguments to the local CLI (@expo/cli).
-npm exec --no-install -- expo-internal "$@"
\ No newline at end of file
diff --git a/bin/cli.js b/bin/cli.js
new file mode 100755
index 0000000000000000000000000000000000000000..1f7e170f29209821c11c89d063dbed0f0c0633aa
--- /dev/null
+++ b/bin/cli.js
@@ -0,0 +1,16 @@
+#!/usr/bin/env node
+const { spawnSync } = require("child_process");
+const path = require("path");
+
+const expoInternalPath = path.join(
+  __dirname,
+  "..",
+  "..",
+  ".bin",
+  "expo-internal"
+);
+// console.log(process.argv.slice(2));
+const result = spawnSync(expoInternalPath, process.argv.slice(2), {
+  stdio: "inherit",
+  cwd: process.cwd(),
+});
diff --git a/package.json b/package.json
index 5dbdf22e586581d3c644aa9fb9a3fa1ff386eb05..d94e4ed569c06b20e3ec39d10f7c9d42536b5022 100644
--- a/package.json
+++ b/package.json
@@ -9,7 +9,7 @@
     "*.fx.js",
     "*.fx.web.js"
   ],
-  "bin": "bin/cli",
+  "bin": "./bin/cli.js",
   "files": [
     "android",
     "bin",
```

**Comment by byCedric** [comment]
Fix #23220 is released in `expo@49.0.0-beta.4`
